README.md

```diff
@@ -347,79 +465,204 @@ This is illustrated below.
+in a request body is a date, at which point it can be converted to a datetime object,
+ready for storing in ICAT. The handler is currently only used in the Python ICAT
+Backend, however this is non-backend specific class.
+
+
+## Exceptions & Flask Error Handling
+Exceptions custom to DataGateway API are defined in `datagateway_api.common.exceptions`.
+Each exception has a status code and a default message (which can be changed when
+raising the exception in code). None of them are backend specific, however some are only
+used in a single backend because their meaning becomes irrelevant anywhere else.
+
+When the API is setup in `main.py`, a custom API object is created (inheriting
+flask_restful's `Api` object) so `handle_error()` can be overridden. A previous
+iteration of the API registered a error handler with the `Api` object, however this
+meant DataGateway API's custom error handling only worked as intended in debug mode (as
+detailed in a
+[GitHub issue](https://github.com/ral-facilities/datagateway-api/issues/147)). This
+solution prevents any exception returning a 500 status code (no matter the defined
+status code in `exceptions.py`) in production mode. This is explained in a
+[Stack Overflow answer](https://stackoverflow.com/a/43534068).
+
+
+## Filtering
+Filters available for use in the API are defined in `datagateway_api.common.filters`.
+These filters are all based from `QueryFilter`, an asbtract class to define any filter
+for the API. Precedence is used to prioritise in which order filters should be applied,
+but is only needed for the Database Backend.
+
+Filtering logic is located in `datagateway_api.common.helpers`.
+`get_filters_from_query_string()` uses the request query parameters to form filters to
+be used within the API. A `QueryFilterFactory` is used to build filters for the correct
+backend and the static method within this class is called in
+`get_filters_from_query_string()`.
+
+
+## Backends
+As described at the top of this file, there are currently two ways that the API
+creates/fetches/updates/deletes data from ICAT. The intention is each backend allows a
+different method to communicate with ICAT, but results in a very similarly behaving
+DataGateway API.
+
+
+### Abstract Backend Class
+The abstract class can be found in `datagateway_api.common.backend` and contains all the
+abstract methods that should be found in a class which implements `Backend`. The typical
+architecture across both backends is that the implemented functions call a helper
+function to process the request and the result of that is returned to the user.
+
+Each backend module contains the following files which offer similar functionality,
+implemented in their own ways:
+- `backend.py` - Implemented version of `datagateway_api.common.backend`
+- `filters.py` - Inherited versions of each filter defined in
+  `datagateway_api.common.filters`
+- `helpers.py` - Helper functions that are called in `backend.py`
+
+
+### Creating a Backend
```

Comment:
So I'm not 100% clear on this part, the documentation seems to imply that the backend will be created for me when running `main.py`, but I could only get it to load the swagger interface when I gave it the pre-prod url (`"ICAT_URL": "https://scigateway-preprod.esc.rl.ac.uk:8181"`). I suppose that I've misunderstood and that the API is created for me, but I would need to have actually created a database (or python backend) myself?

Reply:
The concept of a backend is only a developer concept and isn't something that a user needs to know about. You should've been able to do something like the following though?

```bash
curl --request POST 'http://127.0.0.1:5000/sessions' --header 'Content-Type: application/json' --data '{
"username": "root",
"password": "pw",
"mechanism": "simple"
}'
```

Reply:
Yeah I was able to submit requests through the swagger interface (through browser) or with `curl` after setting `"ICAT_URL": "https://scigateway-preprod.esc.rl.ac.uk:8181"`. My confusion was with whether I'd missed a step and should be say setting something up to run on `localhost:8181` while following the documentation and then provide that in the config file or whether someone following the documentation would already have an ICAT_URL instance set up for them and they only need to modify the config file. I guess from your response it's the latter?

Reply:
I would imagine a 'typical' user would probably have an ICAT_URL already so they only need to modify the config file. As a developer, I have an ICAT instance setup locally which runs on `localhost:8181` (I used some tutorials on the [icat.manual](https://github.com/icatproject/icat.manual/tree/master/tutorials) repo) but any form of prod/pre-prod would use a well-established ICAT URL. I've added a link to these tutorials in the README, as per commit 73bb72d186a321b81bfba4e02484aed37d81e604

Reply:
That makes sense, and I think the extra explanation helps clarify it (though I do appreciate that typical users and experienced devs probably wouldn't need to be explicitly told this).